xds/src/main/java/io/grpc/xds/WeightedTargetLoadBalancerProvider.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.Internal;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancer.Helper;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.LoadBalancerRegistry;
+import io.grpc.NameResolver.ConfigOrError;
+import io.grpc.Status;
+import io.grpc.internal.JsonUtil;
+import io.grpc.internal.ServiceConfigUtil;
+import io.grpc.internal.ServiceConfigUtil.LbConfig;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * The provider for the weighted_target balancing policy.  This class should not be
+ * directly referenced in code.  The policy should be accessed through {@link
+ * LoadBalancerRegistry#getProvider} with the name "weighted_target_experimental".
+ */
+@Internal
+public final class WeightedTargetLoadBalancerProvider extends LoadBalancerProvider {
+
+  static final String WEIGHTED_TARGET_POLICY_NAME = "weighted_target_experimental";
```

Comment:
This is like EDS_POLICY_NAME, which is used by cds lb. CDS_POLICY_NAME will be used by weighted-target-lb. WEIGHTED_TARGET_POLICY_NAME will be used by xds-routing lb. Now only XDS_POLICY_NAME won't be used by other lb.

Reply:
I don't think usages like [`CdsLoadBalancer`](https://github.com/grpc/grpc-java/blob/4a2c5d6e9c0f90de961bc60f2549582e5da1e803/xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java#L295)) is a good practice. `LoadBalancerRegistry` is used for looking up some LB policy given the policy name, the caller treats the policy name opaquely (it just blindly as the registry to give the corresponding LB provider). In CDS LB, it's the design/implementation's choice to load EDS policy. CDS policy explicitly knows what it wants and what that is. It should call `new EdsLoadBalancerProvider()` directly (or even `EdsLoadBalancer(...)` constructor). Existing way of instantiating child policy is an abuse of registry.

Reply:
There are two reasons using register:

- The EDS balancer is a plugin, the built-in implementation can be deregistered and re-registered with a custom implementation plugin with EDS_POLICY_NAME. 

- In testing, we can use a fake EDS balancer for testing CDS.

Reply:
If you look at the javadoc 

```java
/**
   * Register a provider.
   *
   * <p>If the provider's {@link LoadBalancerProvider#isAvailable isAvailable()} returns
   * {@code false}, this method will throw {@link IllegalArgumentException}.
   *
   * <p>If more than one provider with the same {@link LoadBalancerProvider#getPolicyName policy
   * name} are registered, the one with the highest {@link LoadBalancerProvider#getPriority
   * priority} will be effective.  If there are more than one name-sake providers rank the highest
   * priority, the one registered first will be effective.
   */
  public synchronized void register(LoadBalancerProvider provider)
```

Even "round_robin" can be re-registered with different implementations, if multiple round robin lbs are registered, it will pick the one with highest priority.

Reply:
Alright, I accept the current way of looking up registry for creating the downstream EDS policy.